server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -145,6 +145,8 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
                 final ClusterState incomingState;
                 try {
                     incomingState = ClusterState.readFrom(in, transportService.getLocalNode());
+                    // Close early to release resources used by the de-compression as early as possible
+                    in.close();
```

Comment:
We do have compressed mapping sources in the cluster state that itself comes to us in compressed form so not closing here would actually lead to not having nested decompression streams so I added the early close.